### PR TITLE
rebuild stats production env

### PR DESCRIPTION
**Pull request by emmaai**
In light of all the issues we have encountered since C3 update, we have made changes to production environment with more strict and explicit dependency trees.

The issues with `gdal` image, 
- The version of `PROJ` is always the newest, which is unknown when using the image. We're not fully aware of what has been in the new version. It has been difficult for us to keep track.
- All the libs `gdal` depends on are private, the symbol names are changed such that no third-party can use the same set of binaries. It makes it impossible to compile `pyproj` or `cartopy` based on the exactly same `proj` lib, which is the source of trouble when we have been using various of geo python packages. There is no guarantee that they will produce the same results.

The goals we try to achieve,
- Have a very clear dependencies in our production env. It has to be reproducible and trackable when necessary.
- Have consistencies over the production env and test/development env (sandbox). 
- Have consistencies on the final results or products, keep better track on the discrepancies in terms of being alarmed on time and being easy to debug.

Conda is not the only option, but compared with compiling `gdal` from scratch, it's better.

Changes:
- Build new env with `micromamba`
- Upgrade os to ubuntu 22.04 due to lack of compatible rasterio on ubuntu 20.04
- Add docker lint and dive
- Add the data regression test against the image rather than sha
- Push image straight to ECR

### CODE REVIEW

**Comment by jmettes** [comment]
Fantastic work, this is a big improvement! Looks good to me. 🎉

---

In terms of following up on your stated the goals:

> * **Have a very clear dependencies in our production env. It has to be reproducible and trackable when necessary.**

So this is addressed by [pinning the dependencies](https://github.com/opendatacube/datacube-docker/pull/96/files#diff-afcd0513cc478ea2862118192b9c9cb9dbd85cd3455727750a6e7501ac0427faR4) in the conda env yaml?

If one discovered the docker image which produced a product, they could trace the GDAL/proj versions through the `GDAL_DRIVER_PATH`, `PROJ_LIB`, `GDAL_DATA` [environments variables](https://github.com/opendatacube/datacube-docker/pull/96/files#diff-20939c72945cc2d7553671549c3ee4acd8463f4ba9dc21ec9ba78cfdf2cae552R19) set.

I suppose all that's left is recording the docker image in the `.prod-info.yaml` file for the statistician GDAL/proj dependencies.

```
$ aws s3 cp s3://dea-public-data-dev/stats-golden-files/x49/y24/2015--P1Y/ga_ls_tc_pc_cyear_3_x49y24_2015--P1Y_final.proc-info.yaml -
...
software_versions:
- name: eodatasets3
  url: https://github.com/GeoscienceAustralia/eo-datasets
  version: 0.28.1
- name: odc-stats
  url: https://github.com/opendatacube/odc-tools
  version: 1.0.28
- name: ga_tc_percentiles
  url: https://github.com/opendatacube/odc-tools
  version: 1.0.1
- name: datacube-ows
  url: https://github.com/opendatacube/datacube-ows
  version: 1.8.28

...
```
which is non-trivial (but possible) to do with k8s.

Or could [statistician/version.txt](https://github.com/opendatacube/datacube-docker/pull/96/files#diff-494df8e11aabdc25975b772898a514f9bf78b9f900e5b4e8b700df229b4b37e1R58) get written to there?


> * **Have consistencies over the production env and test/development env (sandbox).**

How does Sandbox fit into things - will it reuse this conda environment somehow? Or is it a matter of keeping dependencies in sync? I wonder if it's worth being a [conda package](https://docs.conda.io/projects/conda-build/en/latest/user-guide/tutorials/build-pkgs.html).

> * **Have consistencies on the final results or products, keep better track on the discrepancies in terms of being alarmed on time and being easy to debug.**

Really great work towards this goal!

So this is achieved via `s3://dea-public-data-dev/stats-golden-files/`. My first gut feeling is to version them alongside the code in git, but they seem big enough to be impractical

```
$ aws s3 ls s3://dea-public-data-dev/stats-golden-files/ --recursive --human --summarize
2022-09-20 11:30:16    2.5 KiB stats-golden-files/x49/y24/2015--P1Y/ga_ls8c_nbart_gm_cyear_3_x49y24_2015--P1Y_final.odc-metadata.yaml
2022-09-20 11:30:17    2.1 KiB stats-golden-files/x49/y24/2015--P1Y/ga_ls8c_nbart_gm_cyear_3_x49y24_2015--P1Y_final.proc-info.yaml
2022-09-20 11:30:17    1.4 KiB stats-golden-files/x49/y24/2015--P1Y/ga_ls8c_nbart_gm_cyear_3_x49y24_2015--P1Y_final.sha1
...

Total Objects: 52
   Total Size: 349.4 MiB
```

I can't think of a better way to do it, but it feels fundamental yet hidden away. I suppose the integration test will point to where they are. Perhaps a note in the README? idk, I'm prob splitting hairs.

File: .github/workflows/statistician-image.yml
Comment (SpacemanPaul) [inline .github/workflows/statistician-image.yml:63]: Should this be hard-coded or passed in as an environment variable?

File: statistician/env.yaml
Comment (SpacemanPaul) [inline statistician/env.yaml:9]: I like to try to comment exact version pins where possible, so future developers know what that they need to be aware of when updating.

**Review by SpacemanPaul — APPROVED** [review]
Broadly looks good.  I've added a few minor comments, and I think Jonathan's comments should also be addressed before merging.

File: .github/workflows/statistician-image.yml
Comment (jmettes) [reply .github/workflows/statistician-image.yml:63]: Is it ever going to be different or reused elsewhere? If not, maybe hard-coding is okay?

**Comment by jmettes** [comment]
I know we put in a bunch of work to add ECR syncing... is it worth still keeping in an additional sync to DockerHub? I ask since this is a public opendatacube project. Do we have an obligation to maintain public access to the image, beyond our own use in our cluster?

File: statistician/env.yaml
Comment (emmaai) [reply statistician/env.yaml:9]: The idea here is to build a conda env with the minimum requirements so far we know that it'll be a) working b) without conflict, which is one major benefit of conda, it sorts out the dependencies when it can, and complains when it can't c) reproduce the results of current env we have been using when producing the current summary products, which is guaranteed by the "golden files" stored in aws. 

If we need, we can preserve the env locking list of very specific version infos, including path and sha, after the successful build. Then the same env can be built with the list. I'm not so sure if preserving the list is needed or not. IMHO, the version can be very loose as long as the results/data are the same. It's only for `odc-stats` case though, since any change made in `odc-stats` will be required to pull the latest env and pass the test, the same for the env build. Personally I don't prefer to staying unchanged, but being alarmed when the change introduce the problem and resolve it relatively easily.

File: .github/workflows/statistician-dive.yml
Comment (pindge) [inline .github/workflows/statistician-dive.yml:16]: its called `opendatacube` https://hub.docker.com/r/opendatacube/datacube-statistician

File: .github/workflows/statistician-dive.yml
Comment (pindge) [inline .github/workflows/statistician-dive.yml:17]: and `datacube-statistician` https://hub.docker.com/r/opendatacube/datacube-statistician

File: .github/workflows/statistician-image.yml
Comment (pindge) [inline .github/workflows/statistician-image.yml:63]: put this in secrets,

**Comment by emmaai** [comment]
Re 
@jmettes 

- dependencies are sort of taken care of by conda/mamba. All the libraries/packages share the same set of binaries, they don't carry their own freely. So it's clear enough for us to know that all the packages with similar functionalities have the same source, hence the similar behaviour. If there is something unexpected, to check the package alone is enough. The nightmare we had before, is that, pyproj, catropy and even rasterio, carry their own binary, and versions can vary a lot. Though some part of functionalities provided them are overlapping.
- Sandbox is used as a development env, for validation or even small scale production test run (I used to do it a lot since it has a monster size instance available). So we expect the sandbox and the production env will product the same thing. As I said above, I'm not a believer of staying unchanged or exactly the same. It's unrealistic to some extent, saying we address some issue in the production, it doesn't necessarily an issue in the sandbox. So we won't have to pin the sandbox so closely to the production env, until they produce the different results. conda/mamba is more providing a mechanism of achieving such. To this extent, I do think sandbox or any docker image build should have some test in place, which is naught atm.
- Yes, the size of files is big and hence the reason that they're stored on s3. Gdal can read s3 file easily and check the data. Those are called "golden files" since they're the  "golden standard" of `odc-stats` and production env no matter how they change. It doesn't seem there is any need to version them. Yet agreed that I need to write a more informative readme wrt them.

File: statistician/tests/init_db.sh
Comment (pindge) [inline statistician/tests/init_db.sh:7]: lol? assume this is copied from ows? i can help to do a dump.sql so we don't need to index everytime

**Review by pindge — CHANGES_REQUESTED** [review]

File: .github/workflows/statistician-image.yml
Comment (pindge) [reply .github/workflows/statistician-image.yml:63]: exposing aws account id is aweful

**Comment by emmaai** [comment]
> I know we put in a bunch of work to add ECR syncing... is it worth still keeping in an additional sync to DockerHub? I ask since this is a public opendatacube project. Do we have an obligation to maintain public access to the image, beyond our own use in our cluster?

I don't know...but guess I we can keep pushing to dockerhub if we do.

File: .github/workflows/statistician-dive.yml
Comment (emmaai) [reply .github/workflows/statistician-dive.yml:16]: It only serves as a tag used in local dive, not pushing to anywhere.

File: .github/workflows/statistician-dive.yml
Comment (emmaai) [reply .github/workflows/statistician-dive.yml:17]: comment as above.

File: statistician/tests/init_db.sh
Comment (emmaai) [reply statistician/tests/init_db.sh:7]: The test will be an on-going process. New test/data could be added in the future. I'd prefer to simply indexing more when it requires other than having to dump a database each time. The indexing takes fractional time compared with the whole test.

File: .github/workflows/statistician-image.yml
Comment (emmaai) [reply .github/workflows/statistician-image.yml:63]: all right...though I don't understand why...the whole point of having an assume role and using GitHub openid is to avoid storing the credentials as secrets. So I'll expect exposing account is not a big deal? I could be ignorant, and I'll take advice. Can I have somebody with privilege adding the thing as secret for me then?

Comment (emmaai) [reply .github/workflows/statistician-image.yml:63]: @omad

Comment (omad) [reply .github/workflows/statistician-image.yml:63]: There's been huge amounts of contradictory information in the past about exposing AWS Account IDs. They're sometimes hidden, but, they're exposed in all sorts of places, and are NOT considered sensitive.

As a case in point for this account, we make available public SNS topics for anyone to subscribe to new datsets, and the ARN for SNS Topics includes the account id.

Refs:
- [AWS SNS Notifications — Digital Earth Australia 1.0.0 documentation](https://docs.dea.ga.gov.au/setup/AWS/data_and_metadata.html#notifications)
- [Are AWS account IDs sensitive information? - Last Week in AWS Blog](https://www.lastweekinaws.com/blog/are-aws-account-ids-sensitive-information/)
- [Are AWS Account IDs Sensitive Information? - YouTube](https://www.youtube.com/watch?v=gYJgDymP_GU)

Comment (omad) [reply .github/workflows/statistician-image.yml:63]: I don't think this is necessary, see longer comment above.

**Comment by omad** [comment]
@NikitaGandhi Does DEAfrica pull Datacube Statistician images from DockerHub? And expect them to be updated there?

@emmaai @jmettes @pindge We should decide whether this is going to be a public image that we point ODC users towards, and that we support. OR, we should move this out of being in an ODC GitHub repository and put it in a GeoscienceAustralia repository.

I'm not sure what all the pieces are that it includes, so not sure which way we should go. From my perspective it would definitely be easier to just mange it internally.

I don't think it's been publicised elsewhere, so aside from DEAfrica I'd be very surprised if anyone else was using it.

**Comment by emmaai** [comment]
I haven’t seen any external users activity on odc-stats in the past year, compared with other odc family members. The only purpose of the code is to produce summary products. So I’m guessing it’s more sensible to move it to ga’s repo, manage it internally rather than support a public image.

**Comment by jmettes** [comment]
Looking through DEAfrica's summary stats workflow & infra, it looks like they have a similar workflow to us - syncing `opendatacube/datacube-statistician` to their own ECR.

Also, kind of related: it might also make sense to merge the statistician part of this repo with https://github.com/opendatacube/odc-stats? My guess is that the original intention of this repo was to consolidate Docker configuration between different applications - namely `index` and `statistician`. Perhaps, because the there is an overlap in dependencies (e.g., GDAL) which ought to be synchronised. Does having them in the a central repo make this easier? Otherwise, it might make more sense to keep the Docker configuration alongside the code using it?

**Comment by NikitaGandhi** [comment]
@omad @emmaai DEAfrica uses Datacube `Statistician` and `indexing` images from DockerHub. 
Adding @fangfy and [Seffat Chowdhury](https://github.com/SeffatChowdhury) is looking after DEAfrica data pipelines for any impact and take further action if required.

**Comment by pindge** [comment]
> @omad @emmaai DEAfrica uses Datacube `Statistician` and `indexing` images from DockerHub. Adding @fangfy and [Seffat Chowdhury](https://github.com/SeffatChowdhury) is looking after DEAfrica data pipelines for any impact and take further action if required.

Fang moved to FrontierSI some weeks ago

**Comment by emmaai** [comment]
I don't think the change here will affect deafrica pipeline, the version used by their pipeline is still on dockerhub. As @jmettes suggested, my guess is the dockerhub only serves as transit. I think we can push to deafrica's ecr as well, if we can get some credential injected s we did in ours.

My opinion, it's sensible to gather the docker image and the code together in one repo.

**Comment by emmaai** [comment]
So here are some proposals/options:
- we move `statistician` to be together with `odc-stats`
- we consider pushing the docker image to dockerhub the same time, or
- we open our ecr (DEA's) to public such that i) not denying public access in case anybody using it which I'm not aware ii) not wasting the resource of dockerhub in case truely nobody is using it

**Comment by emmaai** [comment]
close the pr due to new one here https://github.com/opendatacube/odc-stats/pull/66

**Pull request closed without merging**
